Marketplace: Add "Add to Store" button for free and WordPress.org products

raicem opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

- Allows merchants to install any free and .org hosted product from any page, while creating the order on the go. Currently this mocks order creation process.
- Adds new `<InstallNewProductModal>` component. This listens for changes in `query`, and shows the corresponding modal to the user.
- Adds new **Add to Store** button on the search pages. You need to switch to the Woo.com branch as well.

<!-- Begin testing instructions -->


### How to test the changes in this Pull Request:

1. Connect front end and back end to your local Woo.com instance. 
    - For front end, switch `MARKETPLACE_HOST` in `marketplace/components/constants.ts` to `https://woocommerce.test`.
    - For back-end, make modifications in `class-wc-helper`. Make sure you also connect to the right Docker network.
2. Build assets with `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build`
3. Disconnect store from Woo.com by clicking **Avatar > Disconnect** from store
    - Please note which page you initiate the disconnection from. After the disconnect is done, you should be redirected back to the same page.
5. [On the search page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=extensions&path=%2Fextensions), find a product with the **Add to Store** button and click it. It should show a warning for requiring a connection.
6. Complete the connection. Please note when you come back to your store, the modal should open automatically.
7. In the next modal, click Install. This will first ping Woo.com, create an order and then start downloading the product from WordPress.org repository. 
8. When the installation is done and we're allowed to activate (sometimes we're not allowed), you should see the **Activate** button. Clicking the button will activate the product and then show an option to see the plugin list, to set up the plugin. 
9. On Woo.com, delete your customer's billing address and try to install the plugin again. You won't see the **Add to Store** on the same product again, because it's now installed. 
10. Trying to install a new product, should first show a warning and then redirect user to Woo.com with the product added to cart.
11. When it's done, we should test installations with a product key. Find a product key from the API requests in [the My Subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) and then visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions&install=PRODUCT_KEY_STARTS_WITH_W00.
12. On the [the My Subscriptions page](http://localhost:8888/wp-admin/admin.php?page=wc-admin&tab=my-subscriptions&path=%2Fextensions) click **Install** in the table and confirm it works.

raicem committed:
Marketplace: Install free .org plugins with Add to Store button

raicem commented on plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx:
```diff
@@ -0,0 +1,401 @@
+
+				setSubscription( subscriptionData );
+
+				dispatch( installingStore ).startInstalling(
+					subscriptionData.product_key
+				);
```
Do you think I need to dispatch this to the store @bgrgicak?

bgrgicak replied:
Yes, this will keep things consistent with the subscriptions page.
Any reason to not do it?

raicem replied:
No, I asked just to reduce the code as much as possible. I'm happy to leave it in.

raicem commented on plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx:
This file handles the whole flow from start to finish. Therefore, in the end, I didn't need to change anything from the previous installation flows. 

I think file is a little too big. Maybe I'll refactor it down the line, when making tweaks on the feature branch.

raicem commented on plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx:
```diff
@@ -0,0 +1,401 @@
+/**
+ * Internal dependencies
+ */
+import ProductCard from '~/marketplace/components/product-card/product-card';
```
I'm going to visit all files and replace the usage of `~` with the relative imports.

raicem commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx:
```diff
@@ -1,46 +1,92 @@
+	// We hardcode this for now while we only display prices in USD.
+	const currencySymbol = '$';
+
+	if ( product.isInstallable && ! isInstalled ) {
```
I think we need one more check here: seeing this button doesn't guarantee that a user can install it. Installing a plugin depends on the `install_plugins` permission. While we check `manage_woocommerce` to load the page and authenticate APIs, there might a chance that the `install_plugins` cap may not be handed out to the store manager.

bgrgicak replied:
Could we add this check? https://github.com/woocommerce/woocommerce/blob/7068c78101a6062e66d46be0dedfc96ca86102fb/plugins/woocommerce-admin/client/settings-recommendations/recommendations-eligibility-wrapper.tsx#L26

raicem replied:
Oh nice. I'll try this tomorrow and report back

raicem commented on plugins/woocommerce-admin/client/marketplace/components/content/content.tsx:
```diff
@@ -133,6 +134,7 @@ export default function Content(): JSX.Element {
 
 	return (
 		<div className="woocommerce-marketplace__content">
+			<InstallNewProductModal products={ products } />
```
Currently, this flow doesn't work on the Featured/Discover page. Because there is no `Product[]` in the top level. I need to modify the data (it comes in groups and not a flat list), lift it up to `<Content>` and pass it down to the `InstallNewProductModal`. This is going to be a separate PR.

bgrgicak replied:
I would leave this as optional in case we don't have time.

raicem replied:
OK, thanks!

raicem commented on plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx:
```diff
@@ -0,0 +1,401 @@
+			.then( ( response ) => {
+				// This narrows the CreateOrderResponse type to CreateOrderSuccessResponse
+				if ( ! response.success ) {
+					throw response;
```
Throwing this bubbles up the error and it caught by the catch block at the very end. That being said, I think code will never run, because when we're in the `then` block, `response.success` will always be true.

bgrgicak replied:
Why not use `Promise. reject()`?

raicem commented on plugins/woocommerce-admin/client/marketplace/components/install-flow/install-new-product-modal.tsx:
```diff
@@ -0,0 +1,401 @@
+			);
+		}
+
+		if ( installStatus === InstallFlowStatus.installedCannotActivate ) {
```
Sometimes WordPress doesn't allow us to activate the plugin, by omitting the `activateUrl`. It can either:
- Plugin is already active: this happens when plugin is deleted without deactivation and then re-installed. 
- User doesn't have the right capability. 

I guide user to the plugins page in those cases. If you look at the `primaryButton` you'll see the main action is **View Plugins**

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/components/product-card/product-card-footer.tsx:
```diff
@@ -1,46 +1,92 @@
 
 function ProductCardFooter( props: ProductCardFooterProps ) {
-	const { price, currencySymbol, averageRating, reviewsCount } = props;
+	const { product } = props;
```
Would be nice to move this into a component.

raicem replied:
Hmmm, not sure if I understand it. What do you want to move into a component?

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/contexts/types.ts:
```diff
@@ -7,6 +7,7 @@ import { Options } from '@wordpress/notices';
  * Internal dependencies
  */
 import { Subscription } from '../components/my-subscriptions/types';
+import { Product } from '../components/product-list/types';
```
```suggestion
```

bgrgicak commented on plugins/woocommerce-admin/client/marketplace/utils/functions.tsx:
```diff
@@ -234,6 +236,16 @@ function disconnectProduct( subscription: Subscription ): Promise< void > {
 	} );
 }
 
+type WpAjaxReponse = {
```
Should we make these types generally available in Woo? For example here `plugins/woocommerce-admin/client/typings/global.d.ts`?

raicem replied:
`WpAjaxResponseData` is not really complete. Also this calls `window.wp.updates.ajax` and I'm thinking response might vary a lot, depending on the action. Can we type that properly?

bgrgicak commented on plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php:
```diff
@@ -0,0 +1,72 @@
+				'args'                => array(
+					'product_id' => array(
+						'required'          => true,
+						'validate_callback' => function( $argument ) {
+							return is_int( $argument );
+						},
```
```suggestion
						'validate_callback' => 'is_int',
```

raicem replied:
Oddly, this gives me `"Invalid parameter(s): product_id"`. No further info. I guess this doesn't pass the argument into the `is_int` function (just a guess).

bgrgicak commented on plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php:
```diff
@@ -0,0 +1,72 @@
+	exit;
+}
+
+class WC_Helper_Orders_API {
```
Could we extend `WC_Helper_Subscriptions_API` here to reuse `get_permission`?

raicem replied:
I directly called `WC_Helper_Subscriptions_API::get_permission()`. I'd like to avoid extending it.

bgrgicak commented on plugins/woocommerce/includes/admin/helper/class-wc-helper-orders-api.php:
```diff
@@ -0,0 +1,72 @@
+		} catch ( Exception $e ) {
+			return new \WP_REST_Response(
+				array(
+					'message' => 'Could not start the installation process. Reason: ' . $e->getMessage(),
```
```suggestion
					'message' => __(  'Could not start the installation process. Reason: ', 'woocommerce'  ) . $e->getMessage(),
```

bgrgicak commented:
I left some comments and will do a full review soon. All the comments feel optional.

bgrgicak commented:
> For back-end, make modifications in class-wc-helper. Make sure you also connect to the right Docker network.

@raicem could you please clarify this step?

I started testing on`add/19114-order-creation-helper-endpoint` (Woo.com), but wasn't being redirected, so I switched to trunk and it started working.

But now I get a 404 on `/wc/v3/marketplace/create-order`, I will return to testing this tomorrow morning.

raicem committed:
Marketplace: addressed feedback for the the new install free products flow

- Moved notices to the top of the modal
- Updated notice styles slightly
- Updated the CreateOrderSuccessResponse to reflect API changes

raicem commented:
> I started testing onadd/19114-order-creation-helper-endpoint (Woo.com), but wasn't being redirected, so I switched to trunk and it started working.

That's odd, I didn't merge the Woo.com branch, so there is no code to redirect. The redirection is made depending on the `create-order` endpoint response. I'm happy to help you tomorrow morning. 

> could you please clarify this step?
Sorry, I didn't give any details and it sounded cryptic. I meant tweaks we always make to the [$api_base ](https://github.com/woocommerce/woocommerce/blob/23db3eff69ad4606773dfd86e2ce0e06bf4dc874/plugins/woocommerce/includes/admin/helper/class-wc-helper-api.php#L33)and tweaking the request by hooks. I'll send you more details.

raicem committed:
Marketplace: Require the Helper orders API file

bgrgicak reviewed (approved):
This is amazing, thank you @raicem 🚀

raicem committed:
Marketplace: fix linter errors
